### ISSUE TITLE
Some DNS seeders and block explorers don't show Zebra nodes

### ISSUE THREAD

**Issue by teor2345** [ZcashFoundation/zebra#1904]
## Motivation

In #1844, we deployed a few `zcashd` instances on testnet, to make Zebra's testnet tests more reliable. But those nodes aren't showing up in the ECC's or the Foundation's DNS seeders.

As far as we can tell, the nodes are configured correctly, they are synced to tip, their listener ports are open, and they are advertising their external IPv4 addresses. But they don't show up in the DNS of either seeder implementation.

Also, Zebra nodes don't show up in some other DNS seeders or blockchain explorers.

### Bugs to fix

- [x] #4875

### Things to investigate

- [ ] Try pointing a seeder at a long-lived Zebra instance https://github.com/ZcashFoundation/zebra/pull/4870#issuecomment-1204593184

- [ ] `zcashd` seems to ignore peers from `zebrad`: https://github.com/zcash/zcash/issues/3881#issuecomment-1140146874

Other Protocol Mismatches:
- [ ] `version.addr_from` is typically dummy bytes, but Zebra sends the actual details: https://en.bitcoin.it/wiki/Protocol_documentation#version

### List of Block Explorers

Some block explorers show up to 250 nodes.

There are about 150-200 nodes on mainnet that Zebra instances will connect to.
(Recent Zebra versions are limited to 75 outbound and 50 inbound connections, but this is configurable.)

Zebra might connect to fewer nodes because it has strict latency limits.

#### Zebra Nodes

This block explorer shows about 65 nodes, but it includes Zebra:
* https://explorer.zecmate.com/network

#### No Zebra Nodes

This block explorer shows about 250 nodes:
* https://blockchair.com/zcash/nodes (click "Load More" to see all user agents)

These block explorers show connected nodes only:
* https://zcashblockexplorer.com/nodes

Currently down:
* https://explorer.testnet.z.cash/

### Completed

Make Zebra have long-term connections to peers:
- #3120

DNS seeder bugs:
- [x] work out why some DNS seeders return peers with outdated network protocol versions
  - this is a bug in the Foundation's DNS seeder: https://github.com/ZcashFoundation/dnsseeder/issues/10

Metrics:
- [x] #2871

### Optional

We could investigate why DNS seeders ignore `zebrad` nodes:

<details>

Infrastructure:
- [ ] try running `zcashd` on a different network, to see if Google is blocking cryptocurrency traffic
- [ ] try to run a Zebra node with the initial testnet addresses set to some of those `zcashd` nodes, to make sure they respond to requests correctly

Code Review:
- [ ] check the `zcashd` version of our nodes, and the versions accepted by the DNS seeders

</details>

**Comment by str4d** [comment]
This is the [list of conditions](https://github.com/zcash/zcash-seeder/blob/01fc41b063d83a9bc19ebb12e9e48bb7ae659096/db.h#L104-L120) enforced by `zcash-seeder` on peers in order for their IPs to be given out:

- The peer's port must be 8233 on mainnet or 18233 on testnet (DNS only gives out IPs, so the port needs to be predictable).
- The `NODE_NETWORK` service bit must be set.
- The peer's IP must be routable.
- The peer's client version must be at least `REQUIRE_VERSION` (currently 170015 in the repo, 170014 on my own seeder).
- The peer's block height must be at least 0 (we could bump this to require peers to be closer to the tip, but don't currently do so).
- The peer must be reliably reachable (this is implemented by tracking the success or failure of each connection attempt, with an exponential decay of each attempt's weight).

However, before we get to the above checks, the seeder needs to be able to connect to the peer and perform a very minimal handshake and protocol query:
- `version` / `verack`
- `getaddr` / `addr`

**Comment by teor2345** [comment]
I'm pretty sure our mainnet instances satisfy all these criteria.

But they're not configured with their external IP, and Zebra doesn't have external IP auto-detection. So they'll be advertising `0.0.0.0:(default)`. That might be the issue here. (We'd need to implement #1893, or make changes to our deployment configs.)

It's also possible that Google Cloud is filtering cryptocurrency traffic.

**Comment by teor2345** [comment]
We could test by configuring a publicly routable Zebra instance with the correct external IP address.
And ideally outside Google cloud, to eliminate that possibility.

**Comment by mpguerra** [comment]
Could this be related to #794 at all?

**Comment by conradoplg** [comment]
> Could this be related to #794 at all?

I think it's not, the current value seems to follow the rules https://github.com/ZcashFoundation/zebra/blob/8120e8abac29d02c6df2db2c2b5bc666d2c657c0/zebra-network/src/constants.rs#L112
But we can double check.

**Comment by mpguerra** [comment]
#2380 might help us diagnose this one

**Comment by teor2345** [comment]
We could add user-agent metrics to Zebra.
It's a very small change, just copy and tweak the node version metrics.

This would show us if DNS seeders or block explorers are connecting to Zebra.

**Comment by teor2345** [comment]
I spoke to someone who runs a Zebra mainnet node with high uptime.
It looks like their setup should work on DNS seeders and block explorers.

According to the rules posted above: https://github.com/ZcashFoundation/zebra/issues/1904#issuecomment-938212260

So it seems likely that Zebra is implementing some undocumented part of the network protocol differently.
And that's enough to get it rejected from seeders and block explorers.

We might have to do low-level network diagnostics, or look at the detailed logs from a DNS seeder.

**Comment by conradoplg** [comment]
Zecmate is showing Zebra nodes: https://explorer.zecmate.com/network

![image](https://user-images.githubusercontent.com/35766/140075409-92284605-b73b-4126-a2fe-40b707bdec41.png)

**Comment by mpguerra** [comment]
Updated Blockchair/Blockchair.Support#527 to reference zebra nodes seen in zecmate

**Comment by teor2345** [comment]
This is blocked by an investigation of how block explorers handle the Zcash network protocol.

**Comment by teor2345** [comment]
In https://github.com/zcash/zcash/issues/3881#issuecomment-1140146874 , we found out that `zcashd` seems to ignore peers from `zebrad`.

We should try to find out why.

**Comment by softminus** [comment]
> In https://github.com/zcash/zcash/issues/3881#issuecomment-1140146874 , we found out that zcashd seems to ignore peers from zebrad.

To clarify, this was not from `zebrad`, this was from https://github.com/superbaud/zcash-cotyledon/blob/main/src/main.rs; which only answers `Peers` requests.

@str4d said [here](https://github.com/zcash/zcash/issues/3881#issuecomment-1145555036) that

> `zcashd` does not anticipate the existence of network peers that _only_ respond to `getaddr` messages, because that is not how the seeders interact with the network. If such a peer type were desired, it would need a corresponding service bit (to indicate to peers that it does not serve as a regular peer).

so I don't think this *necessarily* reflects a problem with `zebrad`.

**Comment by softminus** [comment]
I ran `rm ~/.zcash/peers.dat; ./src/zcashd -dnsseed=0 -seednode=127.0.0.1 -port=12345` on a machine with a running `zebrad` (not fully synced to chaintip, if that matters) and the `zcashd` was able to get peers from `zebrad`.

**Comment by softminus** [comment]
https://github.com/eqlabs/ziggurat/blob/main/src/tests/conformance/peering.rs#L20-L48 might also be related/helpful

**Comment by teor2345** [comment]
> > In [zcash/zcash#3881 (comment)](https://github.com/zcash/zcash/issues/3881#issuecomment-1140146874) , we found out that zcashd seems to ignore peers from zebrad.
> 
> To clarify, this was not from `zebrad`, this was from https://github.com/superbaud/zcash-cotyledon/blob/main/src/main.rs; which only answers `Peers` requests.

It's still using a `zebra-network` stack, which is one possible cause of the DNS seeder issues.

Thanks for letting us know that `zebrad`'s `Peers` requests work for `zcashd`.

**Comment by teor2345** [comment]
This seems to be fixed for now, if we find issues with a specific block explorer or seeder, we can open another ticket with those details.
